### PR TITLE
BC-589 - BC-630 - Change to reusable workflows

### PR DESCRIPTION
**Pull request by mamutmk5**
# Description
Change to reusable workflows , so that it cann include to other reposetorys.
- this where done for deployment and clean out after branch deletion

## Links to Tickets or other pull requests
https://ticketsystem.hpi-schul-cloud.org/browse/BC-589
https://ticketsystem.hpi-schul-cloud.org/browse/BC-630

## Changes
Change to reusable workflows , so that it cann include to other reposetorys.
- this where done for deployment and clean out after branch deletion

## Approval for review

- [x] All points were discussed with the ticket creator, support-team or product owner. The code upholds all quality guidelines from the PR-template.

> Notice: Please remove the WIP label if the PR is ready to review, otherwise nobody will review it.

### Link to Definiton of Done

More and detailed information on the _definition of done_ can be found [on Confluence](https://docs.schul-cloud.org/pages/viewpage.action?pageId=92831762)

### CODE REVIEW

File: .github/workflows/dispatch.yml [hunk 1]
```diff
@@ -0,0 +1,38 @@
+name: Deploy on Events
+
+on:
+  repository_dispatch:
+    types: [dev-deploy]
+  workflow_dispatch:
+    inputs:
+      sc-branch:
```
Comment (virgilchiriac) [inline .github/workflows/dispatch.yml:8]: would be a good time to name to something else without "sc" 
```suggestion
      branch_name:
```

File: .github/workflows/dispatch.yml [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        required: true
+jobs:
+  branch_name:
+    runs-on: ubuntu-latest
+    outputs:
+      branch: ${{ steps.extract_branch.outputs.branch }}
+    steps:
+      - run: |
+          temp=${{ github.event.client_payload.branch }}
+          git_ref_name=${temp:-'${{ github.event.inputs.sc-branch }}'}
```
Comment (mamutmk5) [inline .github/workflows/dispatch.yml:19]:
```suggestion
          git_ref_name=${temp:-'${{ github.event.inputs.branch_name }}'}
```

File: .github/workflows/dispatch.yml (hunk 1 above)
Comment (mamutmk5) [reply .github/workflows/dispatch.yml]: Add this to an batch with the place where it was used

File: .github/workflows/deploy.yml [hunk 3]
```diff
@@ -24,16 +31,15 @@ jobs:
...
+          echo "git_ref_name=${{ inputs.branch }}" >> $GITHUB_ENV
+          echo git_ref_name ${{ inputs.branch }}
       - uses: actions/checkout@v2
         with:
           repository: hpi-schul-cloud/${{ matrix.repos }}
-          token: ${{ secrets.REPO_ACCESS_TOKEN }}
+          token: ${{ secrets.token }}
           path: ${{ matrix.repos }}
           fetch-depth: 0
       - run: git checkout ${{ env.git_ref_name }}
```
Comment (virgilchiriac) [inline .github/workflows/deploy.yml:45]: if you use arg 
 with:
    ref: $inputs.branch
you don't need to checkout the branch. actions/checkout will handle that for you

Comment (mamutmk5) [reply .github/workflows/deploy.yml:45]: It's better to handle that separately, it's much faster than if it's included in the actions/checkout. 
Especially if the branch not exist.
If actions/checkout try it, it makes three retries for it with 5 to 10 second delay.

File: .github/workflows/clean_workflow.yml [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    if: ${{ always() }}
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        tennens: [ brb, default, nbc ]
+    steps:
+      - run: |
+          echo "git_ref_name=${{ inputs.branch }}" >> $GITHUB_ENV
+          echo git_ref_name ${{ inputs.branch }}
+      - name: SHA Stuf
```
Comment (alriese) [inline .github/workflows/clean_workflow.yml:70]:
```suggestion
      - name: SHA Stuff
```
Maybe we can find a better name here? That's for the prefix gerneration for the domain, right?

**Review by virgilchiriac — APPROVED** [review]

**Pull request merged**